buffer/src/main/java/io/netty/buffer/api/Send.java

```diff
@@ -0,0 +1,129 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer.api;
+
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+/**
+ * A temporary holder of a {@link Resource}, used for transferring the ownership of the
+ * resource from one thread to another.
+ * <p>
+ * Prior to the {@code Send} being created, the originating resource is invalidated, to prevent access while it is being
+ * sent. This means it cannot be accessed, closed, or disposed of, while it is in-flight. Once the resource is
+ * {@linkplain #receive() received}, the new ownership is established.
+ * <p>
+ * Care must be taken to ensure that the resource is always received by some thread.
+ * Failure to do so can result in a resource leak.
+ *
+ * @param <T>
+ */
+public interface Send<T extends Resource<T>> {
+    /**
+     * Construct a {@link Send} based on the given {@link Supplier}. The supplier will be called only once, in the
+     * receiving thread.
+     *
+     * @param concreteObjectType The concrete type of the object being sent. Specifically, the object returned from the
+     *                           {@link Supplier#get()} method must be an instance of this class.
+     * @param supplier           The supplier of the object being sent, which will be called when the object is ready to
+     *                           be received.
+     * @param <T>                The type of object being sent.
+     * @return A {@link Send} which will deliver an object of the given type, from the supplier.
+     */
+    static <T extends Resource<T>> Send<T> sending(Class<T> concreteObjectType, Supplier<? extends T> supplier) {
+        return new Send<T>() {
+            private final AtomicBoolean gate = new AtomicBoolean();
```

Comment:
Are we expecting concurrent access of `Send` instances, why do we need the atomic state guard?

Reply:
Yeah, `Send` is imagined as a way to transfer stuff between threads, so it needs to be thread-safe.

Reply:
Not expected, but they could be. Since buffers are now in an always-owned state, we need `Send` for when we want to enforce ownership transfer. The neat thing about that is the enforcement is via the type system. It would be bad if that could be circumvented with a data race.